src/cruise-ships.js

```diff
@@ -0,0 +1,22 @@
+function Ship(itinerary) {
+    this.itinerary = itinerary;
+    this.currentPort = this.itinerary.ports[0];
+    this.previousPort = null;
+    this.currentPort.addShip(this);
+};
+
+Ship.prototype.setSail = function () {
+    const i = this.itinerary.ports.indexOf(this.currentPort);
+    this.previousPort = this.itinerary.ports[i];
+    this.currentPort.removeShip(this);
+    this.currentPort = null;
+};
+
+Ship.prototype.dock = function () {
+    const indexPreviousPort = this.itinerary.ports.indexOf(this.previousPort);
+    const i = indexPreviousPort + 1;
+    this.currentPort = this.itinerary.ports[i];
```

Comment:
what if you called dock method more then the length of `this.itinerary.ports` array?

Reply:
Resolved this by adding a guard clause to setSail() method, so it can no longer be called if the final port on the itinerary has been reached. This will stop dock() method from being invoked when there are no further ports remaining on the itinerary.
```suggestion
    Ship.prototype.setSail = function () {
  const i = this.itinerary.ports.indexOf(this.currentPort);
  if (i >= this.itinerary.ports.length -1) {
    throw new Error('Itinerary has been completed! Assign a new itinerary to continue.');
  }
  this.previousPort = this.itinerary.ports[i];
  this.currentPort.removeShip(this);
  this.currentPort = null;
};
```

Reply:
Nice work!